File: app/src/main/java/com/dumuzeyn/mp3player/SoundGroupNamer.java
package com.dumuzeyn.mp3player;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/** Builds truthful local names from absolute, human-readable audio characteristics. */
final class SoundGroupNamer {
    private SoundGroupNamer() {
    }

    static ArrayList<SoundGroup> name(List<SoundGroup> source) {
        ArrayList<SoundGroup> ordered = new ArrayList<>(source);
        Collections.sort(ordered, (left, right) -> left.id.compareTo(right.id));
        Set<String> used = new HashSet<>();
        ArrayList<SoundGroup> result = new ArrayList<>();
        for (SoundGroup group : ordered) {
            ArrayList<Descriptor> descriptors = descriptors(group.centroid);
            ArrayList<Descriptor> selected = new ArrayList<>();
            for (Descriptor descriptor : descriptors) {
                if (selected.size() == 2) break;
                if (!containsCategory(selected, descriptor.category)) selected.add(descriptor);
            }
            if (selected.isEmpty()) {
                selected.add(new Descriptor("profile", "Сбалансированный профиль",
                        "Balanced profile", 0.0d));
            }
            String russian = join(selected, true);
            String english = join(selected, false);
            if (used.contains(russian)) {
                for (Descriptor descriptor : descriptors) {
                    if (descriptor.score >= 0.55d
                            && !containsCategory(selected, descriptor.category)) {
                        selected.add(descriptor);
                        russian = join(selected, true);
                        english = join(selected, false);
                        break;
                    }
                }
            }
            used.add(russian);
            result.add(group.named(russian, english));
        }
        return result;
    }

    private static ArrayList<Descriptor> descriptors(double[] values) {
        ArrayList<Descriptor> result = new ArrayList<>();
        double bpm = value(values, TrackAudioProfile.BPM);
        double confidence = value(values, TrackAudioProfile.TEMPO_CONFIDENCE);
        if (confidence >= 0.45d && bpm >= 50.0d) {
            if (bpm < 90.0d) {
                add(result, "tempo", "Медленный темп", "Slow tempo",
                        0.70d + (90.0d - bpm) / 70.0d);
            } else if (bpm < 118.0d) {
                add(result, "tempo", "Умеренный темп", "Moderate tempo", 0.34d);
            } else if (bpm < 145.0d) {
                add(result, "tempo", "Подвижный темп", "Driving tempo",
                        0.55d + (bpm - 118.0d) / 90.0d);
            } else {
                add(result, "tempo", "Быстрый темп", "Fast tempo",
                        0.85d + (bpm - 145.0d) / 70.0d);
            }
        }
        double energy = value(values, TrackAudioProfile.ENERGY);
        if (energy >= 0.22d) {
            add(result, "energy", "Высокая энергия", "High energy",
                    0.65d + (energy - 0.22d) * 3.0d);
        } else if (energy <= 0.075d) {
            add(result, "energy", "Спокойная энергия", "Calm energy",
                    0.58d + (0.075d - energy) * 4.0d);
        }
        double bass = value(values, TrackAudioProfile.BASS);
        if (bass >= 0.34d) {
            add(result, "bass", "Глубокий бас", "Deep bass",
                    0.68d + (bass - 0.34d) * 2.0d);
        } else if (bass <= 0.10d) {
            add(result, "bass", "Лёгкий бас", "Light bass",
                    0.48d + (0.10d - bass) * 2.0d);
        }
        double brightness = value(values, TrackAudioProfile.CENTROID);
        if (brightness >= 0.40d) {
            add(result, "spectrum", "Яркий спектр", "Bright spectrum",
                    0.66d + (brightness - 0.40d) * 2.0d);
        } else if (brightness <= 0.22d) {
            add(result, "spectrum", "Тёплый спектр", "Warm spectrum",
                    0.58d + (0.22d - brightness) * 2.0d);
        }
        double dynamics = value(values, TrackAudioProfile.DYNAMIC_RANGE);
        if (dynamics >= 14.0d) {
            add(result, "dynamics", "Живая динамика", "Expressive dynamics",
                    0.58d + (dynamics - 14.0d) / 20.0d);
        } else if (dynamics > 0.0d && dynamics <= 6.0d) {
            add(result, "dynamics", "Ровная динамика", "Even dynamics",
                    0.55d + (6.0d - dynamics) / 14.0d);
        }
        double rhythm = value(values, TrackAudioProfile.RHYTHM);
        if (rhythm >= 0.16d) {
            add(result, "rhythm", "Чёткий пульс", "Defined pulse",
                    0.56d + (rhythm - 0.16d) * 2.0d);
        } else if (rhythm > 0.0d && rhythm <= 0.055d) {
            add(result, "rhythm", "Плавный пульс", "Smooth pulse", 0.50d);
        }
        Collections.sort(result, (left, right) -> Double.compare(right.score, left.score));
        return result;
    }

    private static void add(ArrayList<Descriptor> target, String category,
            String russian, String english, double score) {
        target.add(new Descriptor(category, russian, english, score));
    }

    private static double value(double[] values, int index) {
        return values != null && index < values.length ? values[index] : 0.0d;
    }

    private static boolean containsCategory(List<Descriptor> values, String category) {
        for (Descriptor value : values) {
            if (value.category.equals(category)) return true;
        }
        return false;
    }

    private static String join(List<Descriptor> values, boolean russian) {
        StringBuilder result = new StringBuilder();
        for (Descriptor value : values) {
            if (result.length() > 0) result.append(" · ");
            result.append(russian ? value.russian : value.english);
        }
        return result.toString();
    }

    private static final class Descriptor {
        final String category;
        final String russian;
        final String english;
        final double score;

        Descriptor(String category, String russian, String english, double score) {
            this.category = category;
            this.russian = russian;
            this.english = english;
            this.score = score;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/SoundGroupNamer.kt
package com.dumuzeyn.mp3player

import kotlin.math.abs

/** Short collection-style names restored from the pre-KMeans implementation. */
internal object SoundGroupNamer {
    private val TRAITS = arrayOf(
        Trait(TrackAudioProfile.ENERGY, "Энергичный поток", "Energetic flow", "Спокойный поток", "Calm flow"),
        Trait(TrackAudioProfile.CENTROID, "Яркий тембр", "Bright timbre", "Тёмный тембр", "Dark timbre"),
        Trait(TrackAudioProfile.BASS, "Глубокий бас", "Deep bass", "Лёгкий бас", "Light bass"),
        Trait(TrackAudioProfile.DYNAMIC_RANGE, "Живой контраст", "Vivid contrast", "Ровный контраст", "Even contrast"),
        Trait(TrackAudioProfile.RHYTHM, "Чёткий пульс", "Crisp pulse", "Плавный пульс", "Smooth pulse"),
        Trait(TrackAudioProfile.TREBLE, "Воздушный верх", "Airy treble", "Тёплый верх", "Warm treble"),
        Trait(TrackAudioProfile.ZERO_CROSSING, "Острый рисунок", "Sharp texture", "Мягкий рисунок", "Soft texture"),
    )
    private val FALLBACK_RU_ADJECTIVES = arrayOf(
        "Лунный", "Чистый", "Глубокий", "Тёплый", "Свежий", "Звонкий", "Мягкий", "Светлый",
    )
    private val FALLBACK_EN_ADJECTIVES = arrayOf(
        "Lunar", "Clear", "Deep", "Warm", "Fresh", "Resonant", "Soft", "Light",
    )
    private val FALLBACK_RU_NOUNS = arrayOf(
        "пульс", "ритм", "поток", "тембр", "контур", "бас", "отклик", "узор",
    )
    private val FALLBACK_EN_NOUNS = arrayOf(
        "pulse", "rhythm", "flow", "timbre", "contour", "bass", "response", "pattern",
    )

    @JvmStatic
    fun name(source: List<SoundGroup>): ArrayList<SoundGroup> {
        val ordered = ArrayList(source)
        ordered.sortWith { left, right -> left.id.compareTo(right.id) }
        val used = HashSet<String>()
        val result = ArrayList<SoundGroup>(ordered.size)
        ordered.forEach { group ->
            val selected = candidates(group.centroid).firstOrNull { used.add(it.russian) }
                ?: fallback(group.id, used).also { used.add(it.russian) }
            result.add(group.named(selected.russian, selected.english))
        }
        return result
    }

    private fun fallback(id: String, used: Set<String>): Candidate {
        val combinations = FALLBACK_RU_ADJECTIVES.size * FALLBACK_RU_NOUNS.size
        val start = (id.hashCode() and Int.MAX_VALUE) % combinations
        for (offset in 0 until combinations) {
            val value = (start + offset) % combinations
            val adjective = value / FALLBACK_RU_NOUNS.size
            val noun = value % FALLBACK_RU_NOUNS.size
            val russian = FALLBACK_RU_ADJECTIVES[adjective] + " " + FALLBACK_RU_NOUNS[noun]
            if (russian !in used) {
                return Candidate(
                    russian,
                    FALLBACK_EN_ADJECTIVES[adjective] + " " + FALLBACK_EN_NOUNS[noun],
                    0.0,
                )
            }
        }
        return Candidate("Звуковой поток", "Sound flow", 0.0)
    }

    private fun candidates(centroid: DoubleArray): ArrayList<Candidate> {
        val result = ArrayList<Candidate>(TRAITS.size)
        TRAITS.forEach { trait ->
            val value = centroid.getOrElse(trait.feature) { 0.0 }
            result.add(
                if (value >= 0.0) {
                    Candidate(trait.highRussian, trait.highEnglish, abs(value))
                } else {
                    Candidate(trait.lowRussian, trait.lowEnglish, abs(value))
                },
            )
        }
        result.sortWith { left, right -> right.score.compareTo(left.score) }
        return result
    }

    private data class Trait(
        val feature: Int,
        val highRussian: String,
        val highEnglish: String,
        val lowRussian: String,
        val lowEnglish: String,
    )

    private data class Candidate(val russian: String, val english: String, val score: Double)
}
